[CL-956] Flexible pages breadcrumbs component

lilkraftwerk opened the pull request:
Basic flexible breadcrumbs component for flexible pages. 

It may need to change to a `onClick` handler rather than a Link depending on the rest of the implementation but that should be a quick fix. Basically I made it simple and straightforward so it's ready to use in building flexible pages but easy enough to change or extend based on our needs once we begin implementing the functionality.

Dead code linter is broken because this component isn't used anywhere yet

![Screen Shot 2022-06-15 at 10 02 24](https://user-images.githubusercontent.com/3614128/173776560-2bbc5722-a406-4478-94e1-7b13b14ebc28.png)
![Screen Shot 2022-06-15 at 10 02 05](https://user-images.githubusercontent.com/3614128/173776564-a98cf2a2-da8a-4f7e-bd45-42fc4e0003c9.png)

lilkraftwerk committed:
Basic breadcrumb component works

lilkraftwerk committed:
Basic breadcrumbs component for Admin pages

lilkraftwerk committed:
Remove unnecessary check for single breadcrumb

kogre committed:
Update source file en.json

kogre committed:
New translations en.json (English)

kogre committed:
Update source file en.json

kogre committed:
New translations en.json (Moroccan Arabic)

kogre committed:
New translations en.json (French, Belgium)

kogre committed:
New translations en.json (Dutch, Belgium)

kogre committed:
New translations en.json (Serbian (Latin))

kogre committed:
New translations en.json (Luxembourgish)

kogre committed:
New translations en.json (English, United Kingdom)

kogre committed:
New translations en.json (English, Canada)

kogre committed:
New translations en.json (Croatian)

kogre committed:
New translations en.json (Spanish, Chile)

kogre committed:
New translations en.json (Portuguese, Brazilian)

kogre committed:
New translations en.json (Serbian (Cyrillic))

kogre committed:
New translations en.json (Polish)

kogre committed:
New translations en.json (Dutch)

kogre committed:
New translations en.json (Italian)

kogre committed:
New translations en.json (Danish)

kogre committed:
New translations en.json (Arabic)

kogre committed:
New translations en.json (Spanish)

kogre committed:
New translations en.json (French)

kogre committed:
New translations en.json (German)

kogre committed:
New translations en.json (German)

kogre committed:
New translations en.json (Swedish)

brentguf committed:
Merge branch 'master' of https://github.com/CitizenLabDotCo/citizenlab into crowdin_master

brentguf committed:
Remove react-aria-live

brentguf committed:
Merge pull request #2190 from CitizenLabDotCo/crowdin_master

New Crowdin updates

brentguf committed:
Merge pull request #2194 from CitizenLabDotCo/CL-997-remove-react-aria-live

Remove react-aria-live

kogre committed:
New translations en.json (Arabic)

IvaKop committed:
Fix admin projects page

Commit:
Translations updated by CI (extract-intl)

IvaKop committed:
Merge pull request #2197 from CitizenLabDotCo/fix-admin-projects-page

Fix admin projects page

IvaKop committed:
Resolve conflict

IvaKop committed:
Merge pull request #2196 from CitizenLabDotCo/crowdin_master

New Crowdin updates

brentguf commented on front/app/components/UI/Breadcrumbs/index.tsx:
```diff
@@ -0,0 +1,94 @@
+        </Text>
+      </Box>
+    );
+  };
```
A SingleBreadcrumb can just be Breadcrumb, right? :-)

brentguf replied:
I would keep a BreadCrumb in 1 file, then import this into the BreadCrumbs file. I would also make the parent container (Breadcrumbs) display: flex, then you won't need display: inline on the children.

And with styles on the parent of a breadcrumb, you shouldn't have to redefine the styles on each component, right?

lilkraftwerk replied:
The whole file is < 100 lines so I don't think it needs to be split, and moving the single breadcrumb to another file doesn't make sense to me as they are never going to be used separately, a single breadcrumb is just a subcomponent of the parent component and won't ever be used in isolation. 

I named it `SingleBreadcrumb` to make it easy to tell apart from `Breadcrumbs`, instead of just one `s` difference between the two.

I can switch it to flex if you'd like but I think functionally it's the same, it's not really a layout component that needs to handle other states, I felt like putting one breadcrumb inline next to the others was fine

brentguf replied:
I'll let [the code speak](https://github.com/CitizenLabDotCo/citizenlab/compare/CL-956-create-breadcrumbs-component...CL-956/review?expand=1). :-) What do you think about the difference?

What I think: I find it easier to understand and there's less repetition (maintenance).
* I do get the argument that this is a mini component so it would be overkill to move the file out here (unless we want to make it more reusable in the future and let the dev pass a styled Breadcrumb as a prop). But then I would just render the component directly in the map and not put it above in a separate component. Much easier to reason about if you don't need to move up the file again, no? https://alexkondov.com/tao-of-react/#avoid-nested-render-functions
* isCurrentPage is a bit of a confusing name to me. isLastBreadcrumb communicates clearer what we want to know (and it shouldn't be part of the props passed to the breadcrumb if I understand its intention well).
* There's also the SingleBreadcrumbWrapper, which could be directly in the code. The conditional part (linkTo) can go inside the component, it's not needed to repeat SingleBreadcrumbWrapper, Divider, etc. But the conditional part isn't even needed, because we should just always require linkTo. This will be easier when creating the data we pas as breadcrumbs props.

Speaking of which, have we thought about how we'll get the data for the breadcrumbs prop? I would make that part of this task as well because we'll need to do that eventually in i1 so could as well be part of this task/PR? If you agree, could you update the ticket to include that info?

lilkraftwerk replied:
I don't really think the old code was difficult to reason about or difficult to maintain, at 92 lines of a purely presentational component. Other than a couple of comments below I think your code is also fine, if you'd prefer to check it in instead please do so.

I don't necessarily agree with that article's point in this context. The argument for having `Divider` defined within `SingleBreadcrumb` is that it doesn't take in props or need to access any variables, it's simply a part of `SingleBreadcrumb` that is needed in two spots in the code, so it's defined in one place and then used as a named component in the two places it's needed.

`isLastBreadcrumb` is a better name for that variable but I believe anyone reading the previous code would have been able to figure out what `isCurrentPage` meant quickly: `isCurrentPage={index === breadcrumbs.length - 1}` is a pretty common convention for checking if something is the final item in an array, and the last breadcrumb generally does represent the "current page" the user is on.

`LinkTo` needs to be optional as the final breadcrumb in a set of breadcrumbs is plain text and not a link, as the link would be to the same page the user is on and therefore kind of useless. In your code the dev is required to supply a link for the final breadcrumb, which is subsequently disabled. I don't think it makes sense to require devs to supply a link that can't be clicked, then disable it/disallow users from clicking on it, rather than just having the final breadcrumb, a non-link with the title of the current page, displayed as text with no link associated.

As for the data, I think it makes sense to wait for someone to use this component in the context of building out the associated feature. Once a dev needs this component they'll be able to see pretty quickly if the current state of entering breadcrumb data matches their needs, and then they can either update it to work for the container's actual needs or use it as-is. 

In general when I code I'm also trying to balance our velocity and our ability to get features out quickly against all of the other tradeoffs. I think my original code was fine for this purpose, as it fulfilled the requirements, was relatively simple and straightforward, and could have been immediately implemented by the dev building the container component for the rest of the feature, which I think is the ultimate goal here. Since it did function as intended, I think trying to fully optimize this component and its code for the purposes of theoretical future maintenance, at this point in the development cycle, is a kind of premature optimization that doesn't necessarily benefit us or our ability to deliver working features.

brentguf replied:
> Other than a couple of comments below I think your code is also fine, if you'd prefer to check it in instead please do so.

Could you copy paste (just 1 file)? Will be easier than merge conflicts.  

> LinkTo needs to be optional as the final breadcrumb in a set of breadcrumbs is plain text and not a link, as the link would be to the same page the user is on and therefore kind of useless. In your code the dev is required to supply a link for the final breadcrumb, which is subsequently disabled. I don't think it makes sense to require devs to supply a link that can't be clicked, then disable it/disallow users from clicking on it, rather than just having the final breadcrumb, a non-link with the title of the current page, displayed as text with no link associated.

I reckon we're going to get them programatically, not manually on each page? So that's why I thought it'll be easier to provide the link to the current page and disable (also with the added benefit it simplifies the component logic). But what you say makes sense, it's weird to pass a link then disable. It'll depend on how we'll fetch the links to previous pages (keep that as simple as possible), so that's why I wanted to make it part of this ticket. It influences how we design the component.  

> As for the data, I think it makes sense to wait for someone to use this component in the context of building out the associated feature. Once a dev needs this component they'll be able to see pretty quickly if the current state of entering breadcrumb data matches their needs, and then they can either update it to work for the container's actual needs or use it as-is.

Usage will be the same everywhere in the admin I believe.

brentguf commented on front/app/components/UI/Breadcrumbs/index.tsx:
```diff
@@ -0,0 +1,94 @@
+const SingleBreadcrumb = ({
+  label,
+  linkTo,
+  isCurrentPage,
```
Optional boolean props should always have a default value (imo). I've been surprised by behavior in the past when I expected a prop wasn't used, then forgot undefined means/converts to false. So I would make it explicit with isCurrentPage = false.

brentguf reviewed (commented):
I'm a bit in the middle, I guess it works, but some changes I'd prefer for code readability/maintainability.

@benfraserdesign I also find the the underline a bit heavy in the UI. How about only showing this underline on hover (if it's not the current page)?

lilkraftwerk committed:
Make boolean prop on breadcrumbs explicit

benfraserdesign commented:
@brentguf sure, happy to remove underlining for cosmetic reasons

IvaKop committed:
[CL-903] Remove terser webpack plugin

IvaKop committed:
[CL-903] Add changelog entry

nTraum committed:
[No ticket] Upgrade to bootsnap 1.12.0

Changelog https://github.com/Shopify/bootsnap/blob/main/CHANGELOG.md

1.12.0
- bootsnap precompile CLI will now also precompile Rakefile and .rake files.
- Stop decorating Module#autoload as it was only useful for supporting Ruby 2.2 and older.
- Remove uname and other patform specific version from the cache keys. RUBY_PLATFORM + RUBY_REVISION should be enough to ensure bytecode compatibility. This should improve caching for alpine based setups. See #409.

IvaKop committed:
Merge pull request #2205 from CitizenLabDotCo/CL-903/fix-safari-crash

[CL-903] Remove terser webpack plugin to fix page crash on Safari

nTraum committed:
Merge pull request #2207 from CitizenLabDotCo/bootsnap

[No ticket] Upgrade to bootsnap 1.12.0

lilkraftwerk committed:
Merge branch 'master' of github.com:CitizenLabDotCo/citizenlab into CL-956-create-breadcrumbs-component

lilkraftwerk committed:
Update breadcrumbs

lilkraftwerk commented:
Ok, I've made the requested changes

lilkraftwerk commented:
New PR here: https://github.com/CitizenLabDotCo/citizenlab/pull/2224
